src/core/types.ts

```diff
@@ -1,3 +1,130 @@
+import {Selection} from 'd3-selection';
+import {BaseMixin} from '../base/base-mixin';
+
+export type ChartParentType = string | BaseMixin | Selection<Element, undefined, Element, unknown>;
+
+export type ColorsList = string[];
+
+export interface MinimalCFDimension {
+    filter(value): this;
+    filterExact(value): this;
+    filterRange(value: any[]): this;
+    filterFunction(value: (k) => any): this;
+    // filterAll(): this; // unused
+    top(k: number): any[];
+    bottom(k: number): any[];
+}
+
+interface CFGrouping {
+    key: any;
+    value: any;
+}
+
+export interface MinimalCFGroup {
+    all: CFGrouping[];
+}
 
 // Used for mix-ins
 export type Constructor<T = {}> = new (...args: any[]) => T;
+
+export type DCPrimitive = number | string | Date;
+
+export type Margins = { top: number; left: number; bottom: number; right: number };
+
+export interface CompareFn {
+    (a: number, b: number): number;
+    (a: string, b: string): number;
+    (a: Date, b: Date): number;
+}
+
+export type GroupingFn = (d) => any;
+
+// Accessors
+
+export type BaseAccessor<T> = (d: any, i?: number) => T;
+export type KeyAccessor = BaseAccessor<any>; // TODO: check if using generics gives better type safety
+export type ValueAccessor = BaseAccessor<any>; // TODO: check if using generics gives better type safety
+export type ColorAccessor = BaseAccessor<any>;
+export type RValueAccessor = BaseAccessor<number>;
+export type LabelAccessor = BaseAccessor<string|number>;
+export type TitleAccessor = BaseAccessor<string|number>;
+
+// Scales
+
+export interface MinimalScaleWithRange<DomainType, RangeType> {
+    (x: DomainType): RangeType;
+
+    domain (): DomainType;
+    domain (newDomain: DomainType): this;
+
+    range (): RangeType[];
+    range (newRange: [RangeType, RangeType]): this;
+}
+
+export type MinimalColorScale = MinimalScaleWithRange<any, string>;
+
+export type MinimalRadiusScale = MinimalScaleWithRange<any, number>;
+
+// Specifically created for scales in Coordinate Grid charts
+export interface MinimalXYScale extends MinimalScaleWithRange<any, number> {
```

Comment:
I wonder if it would make sense to separate the ordinal case from the continuous case, since we use very different parts of the interface. I am only thinking this because this is a lot of `?` to compensate for different usage, not sure if it would actually be helpful.

Reply:
Scales - especially when it comes to CoordinateGrid poses challenges. It is not just Ordinal scales, even `scaleBand` is significantly different. Currently, I have taken an approach that let us do better than what we have. After some more refactoring, we might see if we can use Generic type for x-scale.

No wonder the guys doing typings for d3 have created to no base class for a scale, they have mostly enumerated all the scales.

Once we do generics for the chart data, we can do a little better for scales.

I do realize that Ordinal charts behave differently than others and I am sure this has come up quite few times for discussion.